File: SQLiteMacExtension/with-objC/SQLiteMacExtension.m
#import "SQLiteMacExtension.h"
#import <Foundation/Foundation.h>
#import "sqlite3.h"
//======================================================================================================
//======================================================================================================
static sqlite3 *database;
static sqlite3_stmt *qry;
static NSString *sql_count_rows;
//======================================================================================================
//======================================================================================================
@implementation SQLiteMacExtension

-(NSString *)SQLiteMacGetDocumentsPath{
        NSArray  *paths = NSSearchPathForDirectoriesInDomains (NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *documentsPath = [paths objectAtIndex:0];
        NSLog (@"SQLiteMacGetDocumentsPath: %@", documentsPath);
        return documentsPath;
}

-(NSString *)SQLiteMacOpenDB:(NSString *)db_filename {
        NSString *rslt = @"-1";
        const char *dbpath = [db_filename UTF8String];
        if ( !database ) {
                int rc = sqlite3_open(dbpath, &database);
                if( rc != SQLITE_OK ){
                        rslt = [NSString stringWithFormat:@"SQLiteMacOpenDB:ERROR:%d:%@", rc, [NSString stringWithUTF8String:dbpath]];
                } else {
                        rslt = [NSString stringWithFormat:@"%d", rc];
                }
        }
        return rslt;
}
-(NSString *)SQLiteMacGetLibVersion{
        const char *libversion = sqlite3_libversion();
        const char *sourceid = sqlite3_sourceid();
        NSString *rslt = [NSString stringWithFormat:@"SQLiteMacGetLibVersion:%@:%@:%d",
                [NSString stringWithUTF8String:libversion],
                [NSString stringWithUTF8String:sourceid],
                sqlite3_libversion_number()];
        return rslt;
}
-(NSString *)SQLiteMacCloseDB{
        NSString *rslt = @"-1";
        if ( database ) {
                int rc = sqlite3_close(database);
                database = 0;
                if( rc != SQLITE_OK ){
                        rslt = [NSString stringWithFormat:@"SQLiteMacCloseDB:ERROR:%d", rc];
                } else {
                        rslt = [NSString stringWithFormat:@"%d", rc];
                }
        }
        return rslt;
}
//======================================================================================================
-(NSString *)SQLiteMacExecuteSQL:(NSString *)sql {
        NSString *rslt = @"0";
        char *err_msg;
        int rc = sqlite3_exec(database, [sql UTF8String], NULL, NULL, &err_msg);
        NSLog (@"SQLiteMacExecuteSQL result: %d ", rc);
        if ( rc != SQLITE_OK ) {
                rslt = [NSString stringWithFormat:@"SQLiteMacExecuteSQL:ERROR:%d:%@", rc, [NSString stringWithUTF8String:err_msg]];
        }
        return rslt;
}
//======================================================================================================
-(NSString *)SQLiteMacExecuteInsert:(NSString *)sql {
        NSString *rslt = @"0";
        char *err_msg;
        int rc = sqlite3_exec(database, [sql UTF8String], NULL, NULL, &err_msg);
        NSLog (@"SQLiteMacExecuteInsert result: %d ", rc);
        if ( rc != SQLITE_OK ) {
                rslt = [NSString stringWithFormat:@"SQLiteMacExecuteSQL:ERROR:%d:%@", rc, [NSString stringWithUTF8String:err_msg]];
        } else {
                int row_id = sqlite3_last_insert_rowid(database);
                rslt = [NSString stringWithFormat:@"%d", row_id];
        }
        return rslt;
}
-(NSString *)SQLiteMacExecuteUpdate:(NSString *)sql {
        NSString *rslt = @"0";
        char *err_msg;
        int rc = sqlite3_exec(database, [sql UTF8String], NULL, NULL, &err_msg);
        NSLog (@"SQLiteMacExecuteInsert result: %d ", rc);
        if ( rc != SQLITE_OK ) {
                rslt = [NSString stringWithFormat:@"SQLiteMacExecuteSQL:ERROR:%d:%@", rc, [NSString stringWithUTF8String:err_msg]];
        } else {
                int row_count = sqlite3_changes(database);
                rslt = [NSString stringWithFormat:@"%d", row_count];
        }
        return rslt;
}
-(NSString *)SQLiteMacExecuteDelete:(NSString *)sql {
        NSString *rslt = @"0";
        char *err_msg;
        int rc = sqlite3_exec(database, [sql UTF8String], NULL, NULL, &err_msg);
        NSLog (@"SQLiteMacExecuteInsert result: %d ", rc);
        if ( rc != SQLITE_OK ) {
                rslt = [NSString stringWithFormat:@"SQLiteMacExecuteSQL:ERROR:%d:%@", rc, [NSString stringWithUTF8String:err_msg]];
        } else {
                int row_count = sqlite3_changes(database);
                rslt = [NSString stringWithFormat:@"%d", row_count];
        }
        return rslt;
}
//======================================================================================================
-(NSString *)SQLiteMacQueryOpen:(NSString *)sql {
        NSString *rslt = @"0";
        sql_count_rows = [NSString stringWithFormat:@"SELECT COUNT(*) AS row_count FROM ( %@ )", sql];
        sql_count_rows = [sql_count_rows stringByReplacingOccurrencesOfString:@";" withString:@" "];
        if ( !qry ){
                int rc = sqlite3_prepare_v2(database, [sql UTF8String], -1, &qry, NULL);
                if ( rc != SQLITE_OK ) {
                        const char *msg = sqlite3_errmsg(database);
                        rslt = [NSString stringWithFormat:@"SQLiteMacQueryOpen:ERROR:%d:%@", rc, [NSString stringWithUTF8String:msg]];
                        sqlite3_finalize(qry);
                        qry = 0;
                }
        }
        return rslt;
}
-(NSString *)SQLiteMacQueryGetRowCount {
        NSString *rslt = @"-1";
        sqlite3_stmt *count_qry;
        NSLog (@"SQLiteMacQueryGetRowCount:sql: %@", sql_count_rows);
        int rc = sqlite3_prepare_v2(database, [sql_count_rows UTF8String], -1, &count_qry, NULL);
        if ( rc == SQLITE_OK ) {
                rc = sqlite3_step(count_qry);
                if ( rc == SQLITE_ROW || rc == SQLITE_DONE ){
                        const unsigned char *rslt1 = sqlite3_column_text(count_qry, 0);
                        rslt = [NSString stringWithFormat:@"%s", rslt1];
                }
        }
        sqlite3_finalize(count_qry);
        return rslt;
}
-(NSString *)SQLiteMacQueryGetColumnCount {
        int rc = sqlite3_column_count(qry);
        NSString *rslt = [NSString stringWithFormat:@"%d", rc];
        return rslt;
}
-(NSString *)SQLiteMacQueryGetNext {
        NSString *rslt = @"0";
        int rc = sqlite3_step(qry);
        if ( rc == SQLITE_DONE )
                rslt = @"-1";
        else if ( rc != SQLITE_ROW && rc != SQLITE_DONE ){
                const char *msg = sqlite3_errmsg(database);
                rslt = [NSString stringWithFormat:@"SQLiteMacQueryGetNext:ERROR:%d:%@", rc, [NSString stringWithUTF8String:msg]];
        }
        return rslt;
}
-(NSString *)SQLiteMacQueryGetData:(NSString *)col {
        const unsigned char *rslt1 = sqlite3_column_text(qry, [col intValue]);
                NSString* rslt2 = [[NSString alloc] initWithBytes:rslt1 length:sizeof(rslt1) encoding:NSASCIIStringEncoding];
        return rslt2;
}
-(NSString *)SQLiteMacQueryClose {
        NSString *rslt = @"0";
        if ( qry ){
                int rc = sqlite3_finalize(qry);
                if ( rc != SQLITE_OK ) {
                        const char *msg = sqlite3_errmsg(database);
                        rslt = [NSString stringWithFormat:@"SQLiteMacQueryClose:ERROR:%d:%@", rc, [NSString stringWithUTF8String:msg]];
                }
        }
        qry = 0;
        return rslt;
}
@end
//======================================================================================================
//======================================================================================================

File: SQLiteMacExtension/SQLiteMacExtension.c
#include "sqlite3.h"
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

sqlite3 *db;
sqlite3_stmt *qry = NULL;
char *rslt_str = NULL;
char *sql_count_rows = NULL;
//======================================================================================================
char *SQLiteMacOpenDB(char * db_filename) {
	if ( rslt_str == NULL )
		rslt_str = (char *) malloc(1024);
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	if ( db == NULL ){
		int rc = sqlite3_open( db_filename, &db );
		if ( rc != SQLITE_OK ) {
			sprintf(rslt, "SQLiteMacOpenDB:ERROR:%d:%s:%s", rc, sqlite3_errmsg(db), db_filename);
			sqlite3_close(db);
			db = NULL;
		}
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacGetLibVersion(){
	char * rslt = (char *) malloc(1024);
	sprintf(rslt, "SQLiteMacGetLibVersion:%s:%s:%d", sqlite3_libversion(), sqlite3_sourceid(), sqlite3_libversion_number());
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacCloseDB() {
	char * rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	int rc = sqlite3_close_v2(db);
	if ( rc != SQLITE_OK ) {
		sprintf(rslt, "SQLiteMacCloseDB:ERROR:%d:%s", rc, sqlite3_errmsg(db));	
	}
	db = NULL;
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
//======================================================================================================
char *SQLiteMacExecuteSQL(char *sql){
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	char *err_msg;
	int rc = sqlite3_exec(db, sql, 0, 0, &err_msg);
	if ( rc != SQLITE_OK ) {
		sprintf(rslt, "SQLiteMacExecuteSQL:ERROR:%d:%s", rc, err_msg);
		sqlite3_free(err_msg);
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
//======================================================================================================
char *SQLiteMacExecuteInsert(char *sql){
	char *err_msg = 0;
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	int rc = sqlite3_exec(db, sql, 0, 0, &err_msg);
	if ( rc != SQLITE_OK ) {
		sprintf(rslt, "SQLiteMacExecuteInsert:ERROR:%d:%s", rc, err_msg);
		sqlite3_free(err_msg);
	} else {
		sprintf( rslt, "%lld", sqlite3_last_insert_rowid(db) );
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacExecuteUpdate(char * sql){
	char *err_msg = 0;
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	int rc = sqlite3_exec(db, sql, 0, 0, &err_msg);
	if ( rc != SQLITE_OK ) {
		sprintf(rslt, "SQLiteMacExecuteUpdate:ERROR:%d:%s", rc, err_msg);
		sqlite3_free(err_msg);
	} else {
		sprintf(rslt, "%d", sqlite3_changes(db));
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacExecuteDelete(char * sql){
	char *err_msg = 0;
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	int rc = sqlite3_exec(db, sql, 0, 0, &err_msg);
	if ( rc != SQLITE_OK ) {
		sprintf(rslt, "SQLiteMacExecuteDelete:ERROR:%d:%s", rc, err_msg);
		sqlite3_free(err_msg);
	} else {
		sprintf(rslt, "%d", sqlite3_changes(db));
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
//======================================================================================================
void replaceCharInString(char *s, char old, char new){
	while( *s ) {
		if(*s == old)
		*s = new;
		++s;
	}
}
//======================================================================================================
char *SQLiteMacQueryOpen(char *sql) {
	if ( sql_count_rows == NULL )
		sql_count_rows = (char *) malloc(4096);
	sprintf(sql_count_rows, "SELECT COUNT(*) AS row_count FROM ( %s )", sql);
	replaceCharInString(sql_count_rows, ';', ' ');
	printf("Count Query = %s \n", sql_count_rows);
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	if ( qry == NULL ){
		int rc = sqlite3_prepare_v2(db, sql, -1, &qry, NULL);
		if ( rc != SQLITE_OK ) {
			sprintf(rslt, "SQLiteMacQueryOpen:ERROR:%d:%s", rc, sqlite3_errmsg(db));
			sqlite3_finalize(qry);
			qry = NULL;
		}
	}
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacQueryGetRowCount() {
	sqlite3_stmt *count_qry = NULL;
	strcpy(rslt_str, "-1");
	int rc = sqlite3_prepare_v2(db, sql_count_rows, -1, &count_qry, NULL);
	if ( rc == SQLITE_OK ) {		
		rc = sqlite3_step(count_qry);
		if ( rc == SQLITE_DONE || rc == SQLITE_ROW ){
			strcpy(rslt_str, (char *) sqlite3_column_text(count_qry, 0));
		}
	}	
	sqlite3_finalize(count_qry);
	return rslt_str;
}
char *SQLiteMacQueryGetColumnCount() {
	int rc = sqlite3_column_count(qry);
	sprintf(rslt_str, "%d", rc);
	return rslt_str;
}
char *SQLiteMacQueryGetNext() {
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	int rc = sqlite3_step(qry);
	if ( rc == SQLITE_DONE )
		strcpy(rslt, "-1");	
	else if ( rc != SQLITE_ROW && rc != SQLITE_DONE )
		sprintf(rslt, "SQLiteMacQueryGetNext:ERROR:%d:%s", rc, sqlite3_errmsg(db));
	strcpy(rslt_str, rslt);
	free(rslt);
	return rslt_str;
}
char *SQLiteMacQueryGetData(char * col) {
	return (char *)sqlite3_column_text(qry, atoi(col));
}
char *SQLiteMacQueryClose() {
	char *rslt = (char *) malloc(1024);
	strcpy(rslt, "0");
	if ( qry != NULL ){
		int rc = sqlite3_finalize(qry);
		if ( rc != SQLITE_OK ) {
			sprintf(rslt, "SQLiteMacQueryClose:ERROR:%d:%s", rc, sqlite3_errmsg(db));
		}
	}
	qry = NULL;
	strcpy(rslt_str, rslt);
	free( rslt );
	return rslt_str;
}
//======================================================================================================
// FOR DLL
#if 1
int main() { return 0; }
#else
//======================================================================================================
// FOR EXE
void execQuery(){
	char *rslt = SQLiteMacQueryOpen("SELECT * FROM t1 ORDER BY b;");
	printf("Query open result = %s \n", rslt);
	rslt = SQLiteMacQueryGetRowCount();
	printf("SQLiteMacQueryGetRowCount = %s \n", rslt);
	rslt = SQLiteMacQueryGetColumnCount();
	printf("SQLiteMacQueryGetColumnCount = %s \n", rslt);	
	
	if ( strcmp( rslt, "0" ) == 0 ) {
		int i = 0;
		rslt = SQLiteMacQueryGetNext();
		while ( strcmp( rslt, "0" ) == 0 ) {
			printf("%d: %s - %s \n", i, SQLiteMacQueryGetData("0"), SQLiteMacQueryGetData("1") );
			i++;
			rslt = SQLiteMacQueryGetNext();
		}
		printf("Query end = %s \n", rslt);
		rslt = SQLiteMacQueryClose();
		printf("Query close result = %s \n", rslt);
	}
}

int main() {
	char *rslt = SQLiteMacOpenDB("win.db");
	if ( strcmp( rslt, "0" ) != 0 ) {
		printf("Cannot open database. result is %s \n", rslt);
		return 10;
	}

	printf("Opened database = %s \n", rslt);
	rslt = SQLiteMacExecuteSQL("DROP TABLE IF EXISTS t1;");
	printf("Drop table = %s \n", rslt);
	rslt = SQLiteMacExecuteSQL("CREATE TABLE t1 (a INTEGER PRIMARY KEY, b INTEGER);");
	printf("Created table = %s \n", rslt);
	rslt = SQLiteMacExecuteInsert("INSERT INTO t1 VALUES (1, 123)");
	printf("Insert values = %s \n", rslt);
	rslt = SQLiteMacExecuteInsert("INSERT INTO t1 VALUES (2, 50)");
	printf("Insert values = %s \n", rslt);

	execQuery();
	
	rslt = SQLiteMacExecuteUpdate("UPDATE t1 SET b = 1150 WHERE a = 2;");
	printf("Updated rows = %s \n", rslt);
	
	execQuery();

	rslt = SQLiteMacExecuteDelete("DELETE FROM t1 WHERE a = 1;");
	printf("Deleted rows = %s \n", rslt);
	
	execQuery();

	//rslt = SQLiteMacQueryOpen("SELECT * FROM t1 ORDER BY b;");
	//printf("Query open result = %s \n", rslt);
	
	rslt = SQLiteMacCloseDB();
	printf("Close DB: %s \n", rslt);
	free( rslt );
	return 0;
}
//======================================================================================================
#endif